File: src/game.rs
use geometry::*;
use io::*;
use map::{to_map_pos, to_px_pos};
use pathfind::pathfind;
use states::*;
use world::*;

fn draw_path(px_pos: &P, path: &Vec<P>, io: &mut Io)
{
    for i in 0..path.len() {
        let mut px_pos_prev = if i == 0 {
            *px_pos
        } else {
            to_px_pos(path[i - 1])
        };

        let px_offset = P::new(TILE_SIZE / 2, TILE_SIZE / 2);

        px_pos_prev = px_pos_prev + px_offset;

        let px_pos_current = to_px_pos(path[i]) + px_offset;

        io.draw_line(
            px_pos_prev.x,
            px_pos_prev.y,
            px_pos_current.x,
            px_pos_current.y,
        );
    }
}

// -----------------------------------------------------------------------------
// Game state
// -----------------------------------------------------------------------------
pub struct GameState
{
    world: World,
}

impl GameState
{
    pub fn new() -> GameState
    {
        GameState {
            world: World::new(),
        }
    }
}

impl State for GameState
{
    fn name(&self) -> &str
    {
        return "Game";
    }

    fn on_pushed(&mut self)
    {
    }

    fn on_start(&mut self) -> Vec<StateSignal>
    {
        let mut actor = Actor::new();

        actor.gfx = '@';

        actor.pos = P::new(0, 0);

        actor.time_units = 20;

        self.world.actors.push(actor);

        return Vec::new();
    }

    fn on_resume(&mut self)
    {
    }

    fn draw(&mut self, io: &mut Io)
    {
        // Draw actors
        for actor in self.world.actors.iter() {
            let px_pos = to_px_pos(actor.pos);

            // Draw actor
            io.draw_char(
                actor.gfx,
                px_pos.x + (TILE_SIZE / 2),
                px_pos.y + (TILE_SIZE / 2),
                TextSize::Big,
                TextAnchorX::Mid,
                TextAnchorY::Mid,
            );

            // Draw time units
            io.draw_text(
                &actor.time_units.to_string(),
                px_pos.x + TILE_SIZE,
                px_pos.y,
                TextSize::Small,
                TextAnchorX::Right,
                TextAnchorY::Top,
            );

            // Draw the movement path
            if !actor.movement.path.is_empty() {
                draw_path(&px_pos, &actor.movement.path, io);
            }
        }

        // Draw a grid overlay
        let nr_cells_x = 16;
        let nr_cells_y = 16;

        // Vertical lines
        for x in 0..nr_cells_x {
            io.draw_line(
                x * TILE_SIZE,
                0,
                x * TILE_SIZE,
                nr_cells_y * TILE_SIZE,
            );
        }

        // Horizontal lines
        for y in 0..nr_cells_y {
            io.draw_line(
                0,
                y * TILE_SIZE,
                nr_cells_x * TILE_SIZE,
                y * TILE_SIZE,
            );
        }
    }

    fn update(&mut self, io: &mut Io) -> Vec<StateSignal>
    {
        let blocked_dummy = A2::new_copied(P::new(100, 100), false);

        for actor in self.world.actors.iter_mut() {
            let mov = &mut actor.movement;

            if !mov.is_moving {
                continue;
            }

            // Actor is moving

            assert!(!mov.path.is_empty());

            if actor.time_units <= 0 {
                continue;
            }

            let next_pos = mov.path[0];

            mov.path.remove(0);

            actor.pos = next_pos;

            actor.time_units -= 1;

            // Target reached?
            if mov.path.is_empty() {
                mov.is_moving = false;
            }
        }

        let input = io.read();

        if input.key == Key::Esc {
            return vec![StateSignal::Pop];
        }

        if input.mouse_left_pressed {
            let selected_map_pos = to_map_pos(input.mouse_pos);

            let actor = &mut self.world.actors[0];

            let mov = &mut actor.movement;

            let target_before: Option<P> = if mov.path.is_empty() {
                None
            } else {
                Some(*mov.path.last().unwrap())
            };

            mov.path = pathfind(actor.pos, selected_map_pos, &blocked_dummy);

            if target_before.is_some() {
                if selected_map_pos == target_before.unwrap() {
                    // Selected position is actor's target - start moving
                    mov.is_moving = true;
                }
            }
        }

        return Vec::new();
    }

    fn on_popped(&mut self)
    {
    }
} // impl State for GameState

File: src/game.cpp
#include "game.hpp"

#include "combat.hpp"
#include "component.hpp"
#include "io.hpp"
#include "map.hpp"
#include "pathfinding.hpp"
#include "random.hpp"

// -----------------------------------------------------------------------------
// Private
// -----------------------------------------------------------------------------

// TODO: (Most of) these functions should be moved to other files eventually

static void draw_path(std::vector<P> path, const P& origin)
{
        for (size_t i = 0; i < path.size(); ++i)
        {
                P prev_pos;

                if (i == 0)
                {
                        prev_pos = origin;
                }
                else
                {
                        prev_pos = path[i - 1];
                }

                // TODO: Some repetition of scaling and offsetting here...
                PxPos prev_px_pos;

                prev_px_pos.value =
                        prev_pos.scaled_up(32, 32)
                        .with_x_offset(32 / 2)
                        .with_y_offset(32 / 2);

                const P& current_pos = path[i];

                PxPos current_px_pos;

                current_px_pos.value =
                        current_pos.scaled_up(32, 32)
                        .with_x_offset(32 / 2)
                        .with_y_offset(32 / 2);

                io::draw_line(prev_px_pos, current_px_pos);
        }
}

static void draw_entity(const Entity& e)
{
        if (e.markable && e.markable->is_marked)
        {
                e.markable->draw_mark(*e.pos);
        }

        if (e.movable && !e.movable->path.empty())
        {
                draw_path(e.movable->path, *e.pos);
        }

        e.gfx->draw(*e.pos);
}

static void step_movable(Entity& entity)
{
        if (entity.movable->is_moving)
        {
                entity.pos->set(entity.movable->step());
        }
}

static void step_all_movables(std::vector<Entity>& entities)
{
        for (auto& e : entities)
        {
                if (e.pos && e.movable)
                {
                        step_movable(e);
                }
        }
}

static void unmark_all_entities(std::vector<Entity>& entities)
{
        for (auto& e : entities)
        {
                if (e.markable)
                {
                        e.markable->is_marked = false;
                }
        }
}

static void mark_entities_at(std::vector<Entity>& entities, const P& pos)
{
        for (auto& e : entities)
        {
                if (!e.markable)
                {
                        continue;
                }

                if (e.pos && *e.pos == pos)
                {
                        e.markable->is_marked = true;
                }
        }
}

static void update_movable_for_player_move_order(
        components::Movable& movable,
        const P& pos,
        const P& target,
        const Array2<bool>& blocked)
{
        movable.is_moving = false;

        const auto old_path = movable.path;

        movable.path = pathfind(pos, target, blocked);

        if (movable.path != old_path)
        {
                // We have a new path, which must be confirmed
                // by the player - stop moving
                movable.is_moving = false;

                return;
        }

        if (movable.path.empty())
        {
                return;
        }

        // OK, we have a confirmed, non-empty path

        movable.is_moving = true;
}

static void run_combats(std::vector<Entity>& entities)
{
        for (size_t idx_1 = 0; idx_1 < entities.size(); ++idx_1)
        {
                for (size_t idx_2 = idx_1 + 1; idx_2 < entities.size(); ++idx_2)
                {
                        auto& e1 = entities[idx_1];
                        auto& e2 = entities[idx_2];

                        // TODO: Also check if they are enemies

                        // TODO: If additional armies are here, which are
                        // hostile to both army 1 and 2, they should wait until
                        // army 1 and 2 are completely finished (one of the
                        // armies is killed or has fled), then fight the
                        // winner. There should never be multiple combats going
                        // on at the same position.

                        if (!e1.army ||
                            !e2.army ||
                            !e1.pos ||
                            !e2.pos ||
                            (*e1.pos != *e2.pos))
                        {
                                continue;
                        }

                        // These are two armies at the same position

                        combat::run_combat(
                                e1.army->units,
                                e2.army->units);

                        // TODO: Some duplicated code here, fix
                        for (auto it = begin(e1.army->units);
                             it != end(e1.army->units); )
                        {
                                if (it->strength <= 0)
                                {
                                        it = e1.army->units.erase(it);
                                }
                                else
                                {
                                        ++it;
                                }
                        }

                        for (auto it = begin(e2.army->units);
                             it != end(e2.army->units); )
                        {
                                if (it->strength <= 0)
                                {
                                        it = e2.army->units.erase(it);
                                }
                                else
                                {
                                        ++it;
                                }
                        }

                        if (e1.army->units.empty())
                        {
                                e1.army.reset();

                                e1.discard = true;
                        }

                        if (e2.army->units.empty())
                        {
                                e2.army.reset();

                                e2.discard = true;
                        }

                }
        }
}

static void remove_entities_marked_for_discarding(std::vector<Entity>& entities)
{
        for (auto it = begin(entities); it != end(entities); )
        {
                if (it->discard)
                {
                        entities.erase(it);
                }
                else
                {
                        ++it;
                }
        }
}

// -----------------------------------------------------------------------------
// Game
// -----------------------------------------------------------------------------
std::vector<StateSignal> Game::on_start()
{
        // TODO: These are placeholder sample components
        {
                Entity entity;

                entity.pos = std::make_unique<P>();
                entity.pos->set(8, 8);

                entity.gfx = std::make_unique<components::Gfx>();
                // entity.gfx->gfx = 'M';
                entity.gfx->tile_id = 3;
                entity.gfx->color = {255, 255, 255};

                entity.movable = std::make_unique<components::Movable>();

                entity.markable = std::make_unique<components::Markable>();

                entity.army = std::make_unique<components::Army>();

                entity.army->units.assign(1000, {"foo", 10});

                actors_.push_back(std::move(entity));
        }

        {
                Entity entity;

                entity.pos = std::make_unique<P>();
                entity.pos->set(0, 0);

                entity.gfx = std::make_unique<components::Gfx>();
                entity.gfx->tile_id = 2;
                entity.gfx->color = {255, 255, 255};

                entity.army = std::make_unique<components::Army>();

                entity.army->units.assign(1000, {"foo", 10});

                actors_.push_back(std::move(entity));
        }

        terrain_ = map::generate();

        return {};
}

void Game::draw()
{
        for (auto& e : terrain_)
        {
                draw_entity(e);
        }

        for (auto& e : actors_)
        {
                draw_entity(e);
        }
}

std::vector<StateSignal> Game::update(const InputData& input)
{
        Array2<bool> blocked(terrain_.dims());

        map::update_blocked(terrain_, blocked);

        // TODO: use constants for TILE_SIZE
        // TODO: We need to do something about coordinates! This is stupid.

        if (input.mouse_left_pressed)
        {
                unmark_all_entities(actors_);

                const P selected_map_pos =
                        input.mouse_pos.value.scaled_down(32, 32);

                mark_entities_at(actors_, selected_map_pos);
        }

        if (input.mouse_right_pressed)
        {
                const P selected_map_pos =
                        input.mouse_pos.value.scaled_down(32, 32);

                for (auto& actor : actors_)
                {
                        if (actor.movable &&
                            actor.markable &&
                            actor.markable->is_marked)
                        {
                                update_movable_for_player_move_order(
                                        *actor.movable,
                                        *actor.pos,
                                        selected_map_pos,
                                        blocked);
                        }
                }
        }

        step_all_movables(actors_);

        if (input.c == 'q')
        {
                return {StateSignal().set_pop()};
        }

        // TODO: We don't have any game turn handling yet, so for now we just
        // press "t" to execute a combat for all meeting armies
        if (input.c == 't')
        {
                run_combats(actors_);
        }

        remove_entities_marked_for_discarding(actors_);

        return {};
}
